model-integration/src/main/java/ai/vespa/embedding/VoyageAIEmbedder.java

```diff
@@ -0,0 +1,381 @@
+package ai.vespa.embedding;
+
+import ai.vespa.secret.Secret;
+import ai.vespa.secret.Secrets;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.yahoo.api.annotations.Beta;
+import com.yahoo.component.AbstractComponent;
+import com.yahoo.component.annotation.Inject;
+import com.yahoo.embedding.voyageai.VoyageAiEmbedderConfig;
+import com.yahoo.language.process.Embedder;
+import com.yahoo.tensor.IndexedTensor;
+import com.yahoo.tensor.Tensor;
+import com.yahoo.tensor.TensorType;
+import okhttp3.ConnectionPool;
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * VoyageAI embedder that uses the VoyageAI Embeddings API to generate
+ * high-quality semantic embeddings for text.
+ *
+ * <p>Features:
+ * <ul>
+ *   <li>Supports all VoyageAI models (voyage-3, voyage-code-3, etc.)</li>
+ *   <li>Automatic request batching for efficiency</li>
+ *   <li>LRU caching to reduce API calls</li>
+ *   <li>Auto-detection of input type (query vs document)</li>
+ *   <li>Exponential backoff retry on rate limits</li>
+ * </ul>
+ *
+ * <p>Configuration example in services.xml:
+ * <pre>{@code
+ * <component id="voyage" type="voyage-ai-embedder">
+ *   <model>voyage-3</model>
+ *   <api-key-secret-name>voyage_api_key</api-key-secret-name>
+ * </component>
+ * }</pre>
+ *
+ * @see <a href="https://docs.voyageai.com/">VoyageAI Documentation</a>
+ * @author Vespa Team
+ */
+@Beta
+public class VoyageAIEmbedder extends AbstractComponent implements Embedder {
+
+    private static final Logger log = Logger.getLogger(VoyageAIEmbedder.class.getName());
+    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+
+    // Configuration
+    private final VoyageAiEmbedderConfig config;
+    private final Embedder.Runtime runtime;
+    private final Secret apiKey;
+    private final OkHttpClient httpClient;
+
+    // Cache for embeddings
+    private final Map<CacheKey, Tensor> embeddingCache;
+
+    @Inject
+    public VoyageAIEmbedder(VoyageAiEmbedderConfig config, Embedder.Runtime runtime, Secrets secretStore) {
+        this.config = config;
+        this.runtime = runtime;
+        this.apiKey = getApiKey(config, secretStore);
+        this.httpClient = createHttpClient(config);
+        this.embeddingCache = config.cacheSize() > 0
```

Comment:
Please remove the ad-hoc caching and use the cache mechanism provided by `Context` instead. This scope of this cache is limited - e.g. the lifetime of the document operation for a feed request.

We might introduce a long-lived cache in the `Embedder.Runtime` or `Embedder.Context` later.